Buibaoanh/heb 164 gateway support reasoning config

buibaoanh opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Configurable "reasoning" on model calls (enable, effort levels, optional token limits, exclude).
* **New Models**
  * Added adapters for Claude Opus 4.5, Google Gemini variants, OpenAI GPT OSS variants, and Cohere Embed v4.0.
* **API Changes**
  * POST /chat/completions accepts an optional reasoning field.
* **Refactor**
  * Dynamic adapter system for model/provider selection and provider-specific option mapping.
* **Tests**
  * Expanded tests covering reasoning, model transforms, provider option mappings, and prompt transformations.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

buibaoanh committed:
Support provider-specific reasoning options

buibaoanh committed:
Add configuration for provider-specific reasoning

buibaoanh committed:
Merge remote-tracking branch 'origin/main' into buibaoanh/heb-164-gateway-support-reasoning-config

buibaoanh committed:
feat(reasoning): map reasoning to provider specific options

buibaoanh committed:
fix(reasoning): reasoning.enable should be used as options

buibaoanh committed:
feat(reasoning): Add provider x model reasoning strategies

heiwen commented on apps/gateway/src/utils/converters.ts:
```diff
@@ -16,8 +17,118 @@ import {
+      return {
+        // cannot disable reasoning due to ai sdk lacking include_reasoning option
+        groq: {
+          reasoningEffort: reasoning.effort,
```
What about the other parameters?

buibaoanh committed:
refactor(reasoning): Put reasoning mappings into model config file

buibaoanh committed:
refactor(reasoning): make reasoning configs declarative

buibaoanh committed:
Merge remote-tracking branch 'origin/main' into buibaoanh/heb-164-gateway-support-reasoning-config

buibaoanh committed:
refactor: migrate model config to typescript with reasoning support

buibaoanh committed:
feat(gateway): Enhance provider options handling for reasoning

Refactors how provider-specific options, especially for reasoning, are handled in the gateway:

- **aiModelFactory**: Now returns both the AI SDK model and the `ProviderAdapter` instance, allowing direct access to provider-specific logic.
- **ProviderAdapter**: Introduced `getProviderOptions(reasoning?)` method and `getProviderName()` for polymorphic option key resolution.
- **Bedrock Provider**: Implemented logic to convert reasoning config keys from camelCase to snake_case and wrap them under `additionalModelRequestFields`, addressing Bedrock's API requirements.
- **Vertex Provider**: Overrode `getProviderName` to correctly map to "google" for AI SDK compatibility.
- **Cleanup**: Removed deprecated `toProviderOptions` and related utility functions from `converters.ts`.

buibaoanh committed:
Merge remote-tracking branch 'origin/main' into buibaoanh/heb-164-gateway-support-reasoning-config

buibaoanh committed:
feat(gateway): Implement model and provider adapter pattern

This commit refactors the Gateway application to introduce a robust adapter pattern
for managing AI model configurations and provider interactions. This addresses
the goal of making the Gateway the single source of truth for model definitions
and logic, while preparing for the deprecation of the  package.

Key changes include:

- **Model Adapters:**
  - Introduced  interface and  abstract class for
    standardizing model-specific metadata and configuration transformations.
  - Created dedicated concrete model adapter classes (e.g., ,
    , ) for each supported AI model,
    encapsulating their unique properties (, , ,
    ) and  reasoning transformation logic.
  -  was implemented to provide instances of these specific model adapters.

- **Provider Adapters:**
  - Enhanced  interface and  to include
     and  methods.
  - Individual provider adapters (Bedrock, Cohere, Groq, Vertex) were updated to:
    - Internally define their supported models and their provider-specific IDs,
      removing reliance on .
    - Implement  to declare which model types they can serve.
    - Implement  to convert model-specific configurations into
      their respective provider's API format.

- **Factory Orchestration:**
  -  was refactored to dynamically discover suitable
    provider adapters by iterating through available implementations and querying
    their  method, effectively removing its dependency on the
     file.
  -  was updated to orchestrate the use of
    and  for model creation and validation, returning
    only the AI SDK model instance. Configuration transformations will be applied
    by the consumer of the factory before model invocation.

These changes significantly improve the modularity, extensibility, and testability
of the AI model handling within the Gateway.

buibaoanh committed:
feat(gateway): inject model and provider configuration middlewares

- Wrap chat models with middlewares to transform parameters using model and provider adapters.
- Implement `transformConfigs` in `ProviderAdapterBase` to wrap configs with the provider slug.
- Override `transformConfigs` for Vertex provider to use "google" key.
- Update `providerOptions` in completions module to include reasoning.

buibaoanh committed:
chore(gateway): some clean up on provider adapters

buibaoanh committed:
fix(gateway): resolve type-check errors in ai-model-factory and completions

- Fix LanguageModelMiddleware import and usage
- Resolve implicit any types and missing return types
- Align aiModelFactory.create return type usage
- Adjust providerOptions handling for compatibility with middleware

buibaoanh committed:
chore(gateway): some clean up in adapters

turisanapo commented on packages/shared-data/models/gpt.ts:
```diff
@@ -0,0 +1,56 @@
+  providers: [
+    {
+      bedrock: "openai.gpt-oss-20b-1:0",
+      groq: "openai/gpt-oss-20b",
```
Since each provider is now mapping the exact model id internally, would it make sense to remove the mapping here or is it used somewhere else?

turisanapo commented on apps/gateway/src/middlewares/providers/cohere.ts:
```diff
@@ -11,8 +11,17 @@ export class CohereProviderAdapter
 
-  constructor(modelName: string) {
-    super("cohere", modelName);
+  // modelType to modelId
```
Unnecessary comment.

turisanapo commented on apps/gateway/src/middlewares/providers/groq.ts:
```diff
@@ -11,8 +11,18 @@ export class GroqProviderAdapter
 
-  constructor(modelName: string) {
-    super("groq", modelName);
+  // modelType to modelId
```
Unnecessary comment.

turisanapo commented on apps/gateway/src/middlewares/providers/index.ts:
```diff
@@ -19,25 +18,24 @@ export class ProviderAdapterFactory {
-    for (const providerSlug of providerSlugs) {
-      try {
+    for (const ProviderClass of ALL_PROVIDER_CLASSES) {
+      const tempInstance = new ProviderClass(modelType);
```
Now priority for providers is hardcoded by the list `ALL_PROVIDER_CLASSES` rather than beign delegated to some configuration. Is this intentional?

turisanapo commented on apps/gateway/src/modules/completions.ts:
```diff
@@ -37,6 +39,7 @@ export const completions = new Elysia({
       const toolSet = toToolSet(tools);
       const modelMessages = toModelMessages(messages);
       const coreToolChoice = toToolChoice(toolChoice);
+      const providerOptions = (reasoning ? { reasoning } : {}) as any;
```
Why is `reasoning` renamed to `providerOptions`?

turisanapo commented on packages/shared-data/models/base.ts:
```diff
@@ -0,0 +1,28 @@
+  rateLimit: number;
+  providers: ModelProviderMap[];
+  modality: "chat" | "embedding";
+  family: string;
```
What is `family` used for?

buibaoanh committed:
Merge remote-tracking branch 'origin/main' into buibaoanh/heb-164-gateway-support-reasoning-config

buibaoanh committed:
Refactor(models): Streamline model adapter and factory file structure

- Moved individual model adapter files (cohere, gemini, gpt) from  to  for flatter structure.
- Removed the now-empty  directory.
- Renamed  to  for clarity.
- Renamed  to  to signify its role as the entry point for the models module.
- Updated all affected import paths across the codebase to reflect these changes.

buibaoanh committed:
feat: Rename models.ts to model.ts and update imports

Rename src/middlewares/models/models.ts to src/middlewares/models/model.ts and update all necessary import statements in related files.

buibaoanh committed:
Refactor: Centralize provider model support logic

buibaoanh committed:
Refactor: Use static model support logic in ProviderAdapter

buibaoanh committed:
feat: Enhance ModelAdapter with metadata and rename Gemini adapters

Refactor the `ModelAdapter` interface and `ModelAdapterBase` abstract class to
include `getOwner()`, `getCreatedAt()`, and `getMonthlyFreeTokens()`.

- Renamed `getRateLimit()` to `getMonthlyFreeTokens()` to better reflect
  the quota rather than a rate limit.
- Implemented these new methods in `GptOss120bAdapter`, `GptOss20bAdapter`,
  `GeminiFlashPreviewAdapter`, `GeminiFlashLiteAdapter`, and
  `CohereEmbedV4Adapter`, populating values from `supported-models.json`.
- Renamed `GeminiFlashPreviewAdapter` to `Gemini25FlashPreviewAdapter` and
  `GeminiFlashLiteAdapter` to `Gemini25FlashLitePreviewAdapter` to include
  the model version, preventing future naming conflicts.

buibaoanh committed:
refactor: Remove shared-data and centralize model logic in gateway

- Delete  and all its contents (model definitions, JSON schemas, types).
- Update  to include  and  for model management.
- Implement  in  to replace .
- Update , , and  to remove  dependency.
- Adjust  modules (agents, branches, providers) to use looser types where shared types were removed.

buibaoanh committed:
feat: Dynamically populate model endpoints in Gateway

- Expose  in  to allow iterating over all providers.
- Update  to dynamically generate the  list by checking which providers support the given model.
- Minor cleanup in .

buibaoanh committed:
chore: some comment clean up

buibaoanh committed:
feat(gateway): Centralize options and handle optional configs

- Centralized `OpenAICompatibleOptions` in `src/utils/openai-compatible-api-schemas.ts`.
- Updated models (`cohere`, `gemini`, `gpt`) to use the centralized type and import.
- Implemented `transformConfigs` in `CohereEmbedV4Adapter` to fix type error.
- Updated `transformConfigs` in both `ModelAdapter` and `ProviderAdapter` hierarchies to accept optional inputs (`undefined`) and handle them gracefully.
- Addressed linting issues (unused vars, explicit any) in adapter files.
- Updated bun.lock.

buibaoanh committed:
ci(gateway): fix linter errors

turisanapo commented on apps/api/src/modules/providers.ts:
```diff
@@ -79,20 +78,20 @@ export const providersModule = new Elysia({
       const affectedBranches = branches.filter((branch) => {
-        const models = branch.models as Models;
-        return models.some((model) =>
+        const models = branch.models as any[];
```
It is ok to drop the enum that comes from `supportedModels` but we still want models to be of type `Models`. You can move the type definition on api.

turisanapo commented on apps/gateway/src/middlewares/model-config.ts:
```diff
@@ -1,13 +1,22 @@
+
+import { type SupportedModelType } from "./models";
+
+export interface ModelConfig {
```
If you define models on the API, you can just import from API rather than redifining it here.

buibaoanh committed:
fix(gateway): providerOptions to use transformed params

buibaoanh committed:
refactor(gateway): share Models type between gateway and api

buibaoanh committed:
refactor(api): shared Models type

turisanapo commented on packages/shared-api/types/model-config.ts:
This must be defined using typebox as it was before.

buibaoanh committed:
refactor(api): implement Models with typebox

heiwen commented on apps/gateway/src/modules/models.ts:
```diff
@@ -77,7 +81,8 @@ export const models = new Elysia({
     ({ params }) => {
       const id = `${params.author}/${params.slug}`;
-      const m = supportedModels.find((model) => model.type === id);
+      const supportedModels = getSupportedModels();
```
Nitpick: Inline getSupportedModels() since only used once

heiwen commented on apps/gateway/src/modules/models.ts:
```diff
@@ -25,24 +26,26 @@ const model = t.Object({
-      endpoints: Object.keys(m.providers[0]).map((tag) => ({ tag })),
+      endpoints: ProviderAdapterFactory.ALL_PROVIDER_CLASSES.filter(
+        (ProviderClass) => ProviderClass.supportsModel(m.id),
+      ).map((ProviderClass) => {
```
Nitpick: Should ProviderClass more correctly be ProviderAdapterClass?

heiwen commented on apps/gateway/src/middlewares/providers/index.ts:
```diff
@@ -16,28 +15,26 @@ import { VertexProviderAdapter } from "./vertex";
 import type { ProviderAdapter } from "./provider";
 
 export class ProviderAdapterFactory {
+  static readonly ALL_PROVIDER_CLASSES = [
```
Nitpick: Should this be ALL_PROVIDER_ADAPTER_CLASSES?

heiwen commented on apps/gateway/src/middlewares/providers/provider.ts:
```diff
@@ -2,33 +2,52 @@ import type {
-      | Record<ProviderSlug, string>
-      | undefined;
-    if (!entry) {
+  public getProviderSlug(): ProviderSlug {
```
This feels like it can be a static method?

heiwen commented on apps/gateway/src/middlewares/models/model.ts:
```diff
@@ -0,0 +1,29 @@
+  getMonthlyFreeTokens(): number;
+  getOwner(): string;
+  getCreatedAt(): number;
+  transformConfigs(options?: OpenAICompatibleOptions): Record<string, unknown>;
```
Let's be consistent in naming. Should we use "Options" or "Configs"?

heiwen commented on apps/gateway/src/middlewares/models/index.ts:
```diff
@@ -0,0 +1,61 @@
+
+export type SupportedModel = Static<typeof supportedModel>;
+
+const MODEL_ADAPTER_MAP = {
```
Can you derive this from the Adapters itself, instead of maintaining a second map, same as you do with the Providers?

heiwen commented on apps/gateway/src/middlewares/models/index.ts:
```diff
@@ -0,0 +1,61 @@
+export function getSupportedModels(): SupportedModel[] {
+  return Object.values(MODEL_ADAPTER_MAP).map((factory) => {
+    const adapter = factory();
+    return {
```
Any benefit of all these getters instead of the ModelAdapter returning an object of this shape?

buibaoanh committed:
Merge remote-tracking branch 'origin/main' into buibaoanh/heb-164-gateway-support-reasoning-config

buibaoanh committed:
feat: Add Gemini 3 Pro Preview model support

- Add  to .
- Update  in  to include the new model.
- Update  in  to recognize the new model.

buibaoanh committed:
refactor: update providers to merge nested openai-compatible options

buibaoanh committed:
refactor: simplify model adapter and supported model structure

- Merged `ModelAdapter` and `SupportedModel` types in `src/middlewares/models/model.ts`.
- Replaced getter methods in `ModelAdapter` with readonly properties.
- Refactored all model adapters (`gemini.ts`, `gpt.ts`, `cohere.ts`) to implement the new interface.
- Simplified `src/middlewares/models/index.ts` to use a dynamic `ALL_MODEL_ADAPTER_CLASSES` list.
- Fixed references in `src/modules/models.ts` and `src/middlewares/ai-model-factory.ts`.

buibaoanh committed:
refactor: optimize provider adapter structure

- Convert `providerSlug` to a static property in `ProviderAdapterBase` and subclasses.
- Replace `getProviderSlug()` method with a readonly property in `ProviderAdapter` interface.
- Update `ProviderAdapterFactory` to access static `providerSlug` directly, removing redundant instantiation.
- Remove obsolete `SupportedModelType` alias and update config service to use `string`.

buibaoanh committed:
ci: bun.lock update

turisanapo commented on apps/console/package.json:
```diff
@@ -20,7 +20,7 @@
     "@elysiajs/eden": "^1.4.3",
     "@hebo/aikit-ui": "workspace:*",
-    "@hebo/shared-data": "workspace:*",
+
```
```suggestion
```

buibaoanh commented on apps/gateway/src/middlewares/providers/cohere.ts:
```diff
@@ -11,8 +11,14 @@ export class CohereProviderAdapter
+
+  static readonly SUPPORTED_MODELS_MAP: Record<string, string> = {
+    "cohere/embed-v4.0": "embed-v4.0",
+  };
```
`supportsModel` is a static method inherited by all adapter classes from `ProviderAdapterBase`. It should work.

buibaoanh committed:
Update apps/console/package.json

Co-authored-by: Davide Selvaggio <selvaggiodavide@hotmail.it>

buibaoanh committed:
ci: address code rabbit comment

buibaoanh commented:
Re-testing:
<img width="950" height="629" alt="image" src="https://github.com/user-attachments/assets/44063a6f-8a07-412e-a0c7-e36126f00b72" />

buibaoanh committed:
test: add unit tests for model and provider middleware

Introduces unit tests for the model and provider middleware components to ensure their correct functionality.
Updates  to include a [0m[1mbun test [0m[2mv1.3.3 (274e01c7)[0m script.

buibaoanh committed:
feat(gateway): add gemini 3.0 flash to supported models

buibaoanh committed:
Merge remote-tracking branch 'origin/main' into buibaoanh/heb-164-gateway-support-reasoning-config

# Conflicts:
#	packages/shared-data/json/supported-models.json

buibaoanh committed:
fix(gateway): standardize provider option transformation

Ensure empty options return empty object and preserve sibling provider options during transformation.

buibaoanh commented:
Retest:
<img width="920" height="884" alt="image" src="https://github.com/user-attachments/assets/6aa98996-de50-4fc1-b8e3-2c7945cc6cb6" />
<img width="383" height="551" alt="image" src="https://github.com/user-attachments/assets/1e74fe5c-5e5c-4986-bacf-83d9a2fbb910" />

buibaoanh committed:
feat(gemini): support thinkingLevel for Gemini 3 models

- Refactor GeminiModelAdapter to use Template Method pattern
- Implement thinkingLevel for Gemini 3 (Pro defaults HIGH, Flash defaults MEDIUM)
- Retain thinkingBudget for Gemini 2.5
- Add tests for all scenarios

buibaoanh committed:
fix: thinkingLevel value should be lowercase

buibaoanh committed:
Merge branch 'main' into buibaoanh/heb-164-gateway-support-reasoning-config

buibaoanh committed:
test: fix typo on excluded

buibaoanh committed:
fix: preserve additional provider options in model adapters

Update transformOptions in Cohere, Gemini, and GPT adapters to explicitly preserve and propagate non-OpenAI-compatible options.

buibaoanh committed:
refactor: optimize transformOptions and fix tests

Update Gemini and GPT model adapters to use early return pattern for openai-compatible options, aligning with ProviderAdapterBase implementation. Update tests to match the new behavior where empty options return empty object.

buibaoanh committed:
test: add some test for no options

buibaoanh committed:
Merge branch 'buibaoanh/heb-164-gateway-support-reasoning-config' of github.com:heboai/hebo into buibaoanh/heb-164-gateway-support-reasoning-config

buibaoanh committed:
Revert "test: fix typo on excluded"

This reverts commit 7e869bcaa1af7ec45aa1375aa3c9db9e93dc0261.

buibaoanh committed:
ci: fix type check error

buibaoanh committed:
ci: fix linter erro

buibaoanh commented on apps/gateway/tests/middlewares/models/model.test.ts:
```diff
@@ -0,0 +1,330 @@
+          },
+        },
+      },
+    },
```
can you verify again, I cannot find any `excluded` in the code

buibaoanh committed:
feat(gateway): support new reasoning effort levels (none, minimal, xhigh)

Adds 'none', 'minimal', and 'xhigh' to OpenAICompatibleReasoning schema. Updates Gemini adapters to map these new levels to appropriate thinking budgets and levels. Includes comprehensive tests.

buibaoanh committed:
fix: inject google thoughtSignature dummy value

buibaoanh committed:
refactor(gateway): normalize openai-compatible options to camelCase and inject thought signature

- Rename 'openai-compatible' config key to 'openaiCompatible' across all adapters
- Add Gemini3ModelAdapter with thought signature injection for tool calls
- Update converter utilities to handle nested provider options

buibaoanh committed:
fix(gateway): pass provider metadata in tool calls and remove logs

- Update  to merge  into tool calls.
- Remove debug console logs in  and .
- Apply formatting changes to root .

heiwen commented on apps/gateway/src/middlewares/models/cohere.ts:
```diff
@@ -0,0 +1,25 @@
+  };
+
+  transformOptions(options?: ProviderOptions): ProviderOptions {
+    const { openaiCompatible: openAiOptions, ...rest } = options || {};
```
1/ To check whether this works for both "OpenAI" and "OpenAICompatible" provider options?
2/ Since there is no actual transform happening, should there just be a default in the base class?

buibaoanh replied:
1. The options is for OpenAICompatible, let me rename to avoid confusion
2. Yes, this can be moved to base class

buibaoanh commented on apps/gateway/src/middlewares/models/gemini.ts:
```diff
@@ -0,0 +1,218 @@
+      }
+      return message;
+    });
+  }
```
it is an array, from github:
`export type LanguageModelV3Prompt = Array<LanguageModelV3Message>;`

buibaoanh commented on apps/gateway/src/utils/converters.ts:
```diff
@@ -105,12 +105,30 @@ export function toModelMessages(messages: OpenAICompatibleMessage[]) {
+              }
+
+              return content;
+            }),
```
I don't think this is needed, it is wrong to silence the error and worst if rethrow accidentally causing the misleading.
My PR has nothing to do with current logic of input.

buibaoanh committed:
chore: fix some type issues

buibaoanh commented:
Testing:
Receiving thought signature:
<img width="970" height="676" alt="image" src="https://github.com/user-attachments/assets/9cc47df7-c827-4e64-86e4-84d668d29540" />
Pass through thought signature:
<img width="954" height="620" alt="image" src="https://github.com/user-attachments/assets/8b387815-fd05-49ee-9360-13ec55023a2c" />
Inject dummy thought signature when not available:
<img width="593" height="498" alt="image" src="https://github.com/user-attachments/assets/5e3c6e41-df8e-46ca-8a0e-e925e766e15b" />

heiwen commented on apps/gateway/src/middlewares/models/gemini.ts:
```diff
@@ -0,0 +1,218 @@
+                return part;
+              }
+
+              const existingSignature = (part as any).providerOptions
```
I don't think the signature is part of the providerOptions, is it? I thought it's part of the message parts instead?

buibaoanh replied:
yes, it is part of message (tool call), but in separate field,  and since this is provider-specific (google), we use `providerOptions` for injection.

heiwen commented on apps/gateway/src/middlewares/models/gemini.ts:
```diff
@@ -0,0 +1,218 @@
+    const thinkingConfig: Record<string, any> = {};
+
+    if (params.max_tokens === undefined) {
+      switch (params.effort) {
```
What about "medium"?

heiwen commented on apps/gateway/src/utils/converters.ts:
```diff
@@ -105,12 +105,30 @@ export function toModelMessages(messages: OpenAICompatibleMessage[]) {
-            })),
+            content: message.tool_calls.map((toolCall) => {
+              const {
+                // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Which var is unused?

heiwen commented on apps/gateway/src/utils/converters.ts:
```diff
@@ -105,12 +105,30 @@ export function toModelMessages(messages: OpenAICompatibleMessage[]) {
+                input: JSON.parse(toolFunction.arguments),
+              };
+
+              if (Object.keys(metadata).length > 0) {
```
Why are you moving the metadata from the content part into the providerOptions? Isn't the metadata part-specific and the provider options are call-level?

buibaoanh replied:
AI SDK does accept providerOptions at content-part level, which is not being mentioned in the documentation. 

FYR: https://github.com/vercel/ai/blob/main/packages/ai/src/prompt/content-part.ts#L82

heiwen commented on apps/gateway/src/utils/converters.ts:
```diff
@@ -170,14 +188,24 @@ export const toOpenAICompatibleMessage = (
-    }));
+    message.tool_calls = result.toolCalls.map((toolCall: any) => {
+      const metadata = {};
+      if (toolCall.providerMetadata) {
```
Is providerMetadata the same as providerOptions above or is that different?

buibaoanh replied:
`providerOptions` is for input to generateText. However, for the results, `providerMetadata` represents the output.

heiwen commented on apps/gateway/src/utils/openai-compatible-api-schemas.ts:
```diff
@@ -122,4 +144,23 @@ export type OpenAICompatibleToolCallDelta = {
+  [key: string]: any;
 };
+
+export const OpenAICompatibleConfig = t.Object(
```
Are OpenAICompatibleConfig and OpenAICompatibleOptions ever used?

turisanapo commented on apps/gateway/src/middlewares/models/gpt.ts:
```diff
@@ -0,0 +1,74 @@
+      throw new Error("GPT models do not support 'max_tokens' for reasoning.");
+    }
+    if (params.exclude !== undefined) {
+      throw new Error("GPT models do not support 'exclude' for reasoning.");
```
Shall we catch this and let the gateway return a 400? You can use `BadRequestError` upstream.

turisanapo commented on apps/gateway/src/middlewares/providers/index.ts:
```diff
@@ -16,28 +15,27 @@ import { VertexProviderAdapter } from "./vertex";
       }
     }
+
     throw new Error(
```
Shall we catch this and let the gateway return a 400? You can use `BadRequestError` upstream.

buibaoanh committed:
fix(gateway): address H comments

buibaoanh committed:
fix(gateway): address Davide comments

buibaoanh commented on apps/gateway/src/middlewares/models/gemini.ts:
```diff
@@ -0,0 +1,220 @@
+                  google: {
+                    ...(part as any).providerOptions?.google,
+                    thoughtSignature: "context_engineering_is_the_way_to_go",
+                  },
```
openaiCompatible is expected in transformPrompt method because input is generic, the purpose of transformation logic for adapter classes is to make it compatible to specific model / provider.

buibaoanh committed:
fix(gateway): address Davide comments

buibaoanh committed:
ci(gateway): unit tests

buibaoanh committed:
feat(gateway): add Claude Opus 4.5

buibaoanh committed:
feat(gateway): add reasoning config for  Claude Opus 4.5

buibaoanh committed:
refactor(gateway): remove support for Gemini 2.5 models

buibaoanh committed:
test(gateway): restructure model tests

buibaoanh committed:
fix(gateway): maintain correct provider name when existing signature is present

buibaoanh committed:
fix(gateway): should throw exception when max_tokens is set for gemini 3 models

turisanapo commented on apps/gateway/src/middlewares/models/claude.ts:
```diff
@@ -0,0 +1,107 @@
+      thinkingConfig.budgetTokens = params.max_tokens;
+    }
+
+    if (thinkingConfig.budgetTokens < 1024) {
```
Does this mean that thinking is always enabled for Claude Opus 4.5?

turisanapo commented on apps/gateway/src/middlewares/providers/index.ts:
```diff
@@ -16,28 +16,28 @@ import { VertexProviderAdapter } from "./vertex";
 
 export class ProviderAdapterFactory {
+  static readonly ALL_PROVIDER_ADAPTER_CLASSES = [
+    // FUTURE: error-prone for provider fallback feature
```
```suggestion
    // FUTURE: error-prone for provider fallback feature. Fallback provider depends on the order of this list when the same model is supported by multiple providers.
```

buibaoanh committed:
Update apps/gateway/src/middlewares/providers/index.ts

Co-authored-by: Davide Selvaggio <selvaggiodavide@hotmail.it>

buibaoanh committed:
fix(gateway): use extra_content for providerMetadata

buibaoanh committed:
test(gateway): model specific options should not be under openaiCompatible

buibaoanh committed:
test(gateway): add complete reasoning testing for providerOptions transformations

buibaoanh committed:
Merge branch 'buibaoanh/heb-164-gateway-support-reasoning-config' of github.com:heboai/hebo into buibaoanh/heb-164-gateway-support-reasoning-config

buibaoanh commented on apps/gateway/tests/middlewares/models/gemini.test.ts:
```diff
@@ -0,0 +1,624 @@
+            thinkingLevel: "high",
+          },
+        },
+      },
```
please relook, I have addressed this issue

buibaoanh commented on apps/gateway/tests/middlewares/models/gemini.test.ts:
```diff
@@ -0,0 +1,624 @@
+          },
+        },
+      },
+    },
```
please relook, I have addressed this issue
